Sources/Valet/Internal/Keychain.swift

```diff
@@ -48,216 +48,173 @@ internal final class Keychain {
             var secItemQuery = attributes
             secItemQuery[kSecAttrAccount as String] = canaryKey
             secItemQuery[kSecValueData as String] = Data(canaryValue.utf8)
-            _ = SecItem.add(attributes: secItemQuery)
+            try? SecItem.add(attributes: secItemQuery)
             
             return isCanaryValueInKeychain()
         }
     }
     
     // MARK: Getters
     
-    internal static func string(forKey key: String, options: [String : AnyHashable]) -> SecItem.DataResult<String> {
-        switch object(forKey: key, options: options) {
-        case let .success(data):
-            if let string = String(data: data, encoding: .utf8) {
-                return SecItem.DataResult.success(string)
-            } else {
-                return SecItem.DataResult.error(errSecItemNotFound)
-            }
-        case let .error(status):
-            return SecItem.DataResult.error(status)
+    internal static func string(forKey key: String, options: [String : AnyHashable]) throws -> String {
+        let data = try object(forKey: key, options: options)
+        if let string = String(data: data, encoding: .utf8) {
+            return string
+        } else {
+            throw KeychainError.itemNotFound
         }
     }
     
-    internal static func object(forKey key: String, options: [String : AnyHashable]) -> SecItem.DataResult<Data> {
+    internal static func object(forKey key: String, options: [String : AnyHashable]) throws -> Data {
```

Comment:
What do you think about returning `nil` here when the item is not found, rather than throwing `.itemNotFound`? Feels like throws is more about something going wrong / invalid configuration, where an item not existing yet is a valid case in normal usage.

Reply:
I like that a lot. I may need to re-create the `noItemsFound` error for the migration case. If I can treat `itemNotFound` consistently everywhere (outside of migration) with nullability, I agree that’s more semantic. Good idea!

Reply:
Addressed in 873cda1

Reply:
So now that I’ve done it, I’m not convinced I like it. Combining throws and nullability means that accessing a value requires handling two forms of errors. Handling an error in addition to testing for optionality becomes:
```
do {
    guard let object = try valet.objectForKey(“key”) else {
        // handle item not found
        return
    }
    // do thing with object
} catch someErrorWeCareAbout {
    // handle this error
} catch {
    // Nothing to do here, we don’t care about this case
}
```

If we consider “item not found” an error like the other possible errors, the call site code-flow is simpler (no nested indents) since we’re handling all error cases with the same syntax:
```
do {
    let object = try valet.objectForKey(“key”)
    // do thing with object
} catch itemNotFound {
    // handle item not found
} catch someErrorWeCareAbout {
    // handle this error
} catch {
    // Nothing to do here, we don’t care about this case
}
```

Meanwhile, determining “could we find something” reads the same in both the only-error and optional+error approaches:
```
if let object = try? valet.objectForKey(“key”) {
    // do thing with object
}  else {
    // handle item not found
}
```

I’m considering reverting 873cda1. Curious what you think!

Reply:
Ah interesting. Yeah, after seeing the syntax at the call site I think you're right. In theory, it feels like they should be handled differently; but in practice, I think throwing an error makes for cleaner usage.

Reply:
Yeah I'm there with you on the theory; that's what led me to make the change. When reading the tests having to catch just to find `itemNotFound` is also a bit cumbersome. But in practice, I'm guessing folks who care about `itemNotFound` will first try `containsObject(forKey:)`.

The more I think about this, the less convinced I am that `containsObject(forKey:)` is a useful API on a regular `Valet` when `object(forKey:)` and `string(forKey:)` both `throw` specific, catchable errors. Why `try` `containsObject(forKey:)` when you could try accessing the object or string and handle `itemNotFound` in a `catch`? (Note that this API _is_ useful on a `SecureEnclave` or `SinglePromptSecureEnclave` Valet, since accessing the data has a user-prompt side effect and `containsObject` does not.).

Thoughts on whether we should keep `containsObject(forKey:) throws -> Bool` on `Valet`?

In any case, I've reverted 873cda1. @fdiaz curious to know your thoughts on this thread as well.

Reply:
My gut reaction is to keep `containsObject(forKey:)` on a regular `Valet`. Partly because this keeps the API of the various types of Valets more similar, so you can use the same patterns everywhere.

Totally agreed it doesn't add much when you're simply checking for the presence of a key.
```swift
let hasKey = (try? valet.containsObject(forKey: "key") == true)

let hasKey = (try? valet.object(forKey: "key") != nil)
```

There is still some utility if you're checking for the _absence_ of a key though, since otherwise you need to catch the error to see if it's `.itemNotFound`. To expand on this, say you want to change logic in your method based on the presence of a key, and propagate any errors upwards. With `containsObject(forKey:)` you can call it directly with a `try`, vs. calling `object(forKey:)` in a try/catch and handling one case explicitly.

Reply:
Fair point. I guess if you don't want to catch all errors within a method, it's entirely valid to:
```
if try valet.containsObject(forKey: "key") {
    let value = try valet.object(forKey: "key")
    // Do things with value
} else {
    // Do things to prompt the user for a value
}
```

Certainly simpler than:
```
do {
    let value = try valet.object(forKey: "key")
    // Do things with value
} catch KeychainError.itemNotFound {
    // Do things to prompt the user for a value
} catch { error in
    throw error
}
```

Good thinking @NickEntin. Thanks for helping me think through this.